library/Core/src/Juvix/Core/Parameterisation.hs

```diff
@@ -32,3 +33,44 @@ data Parameterisation primTy primVal
         floatVal :: Double -> Maybe primVal
       }
   deriving (Generic)
+
+{-# DEPRECATED
+  parseTy,
+  parseVal,
+  reservedNames,
+  reservedOpNames
+  "TODO: update parser to not use these"
+  #-}
+
+data ApplyError' e a
+  = ExtraArguments a (NonEmpty a)
+  | InvalidArguments a (NonEmpty a)
+  | Extra e
+  deriving (Eq, Show, Functor)
+
+type ApplyError a = ApplyError' (ApplyErrorExtra a) a
+
+class CanApply a where
+  type ApplyErrorExtra a
+  type ApplyErrorExtra a = Void
```

Comment:
What does the double type declaration do here

Reply:
The one with an RHS is a default, so if you don't have any special type errors you can just omit it entirely